### ISSUE TITLE
Completion for non-Invoke programs

### ISSUE THREAD

**Issue by bitprophet** [pyinvoke/invoke#301]
I.e. apps like Fabric v2 which leverage `invoke.program.Program`.

Presently, my local dev workstation has an item in its dotfiles that says `source ~/Code/oss/invoke/completion/zsh`. Works great for `inv`. Doesn't do jack for `fab`.

The upshot of that Invoke completion script is that it looks to `invoke --complete -- <whatever is being tab completed>`. If I do the same for `fab` (with a local `fabfile.py` - without any `fabfile.py`, or `tasks.py` for invoke, things don't work at all -- not great -- needs to be another ticket) it works as well.

So the question is "how should developers using Invoke to drive their own non-`inv` CLI apps, distribute tab completion helper scripts?". Not sure what works great here - some silly meta-script that generates files like `/completion/zsh` with the correct per-project binary name(s) substituted?

**Comment by nhoening** [comment]
I think this is actually crucial for everyone building their own apps. I'm doing this for an in-house tool right now and I love Invoke for this. It took me as a bash/zsh amateur a while but in the end the actual changes are just two little replacements. For the record, here are my changes to the zsh and bash completion scripts (maybe someone can use these until this issue gets resolved):

zsh:

```
20c20,21
<     reply=( $(invoke ${=collection_arg} --complete -- ${words}) )
---
>     # with a custom app name, `invoke --complete` chokes when working on options
>     reply=( $(mytool ${=collection_arg} --complete -- ${words//mytool/invoke}) )
24c25
< # Tell shell builtin to use the above for completing 'inv'/'invoke':
---
> # Tell shell builtin to use the above for completing 'mytool':
30c31
< compctl -K _complete_invoke + -f inv invoke
---
> compctl -K _complete_invoke + -f mytool
```

bash:

```
11c11,12
<     candidates=`invoke --complete -- ${COMP_WORDS[*]}`
---
>     # with a custom app name, `invoke --complete` chokes when working on options
>     candidates=`mytool --complete -- "${COMP_WORDS[*]/mytool/invoke}"`
26c27
< # Tell shell builtin to use the above for completing 'inv'/'invoke':
---
> # Tell shell builtin to use the above for completing 'mytool':
30c31
< complete -F _complete_invoke -o default invoke inv 
---
> complete -F _complete_invoke -o default mytool
```

Let me know if this is done wrong but I found I had to make these replacements, so that `invoke --complete` would work as intended.

As for an actual solution for this issue, two quick thoughts:

* As you said, provide a "silly" customisation script which applies a patch similar to the above and which accepts the custom app name as a parameter (e.g. "myapp" for the example above). Why not? Invoke already is supporting the completion in an ad-hoc fashion ("source this if you please"), this would just add an intermediate step.
* An attempt at an elegant solution might be applicable for the case where someone writes code that leverages `invoke.program.Program`. `Program` could accept a configuration option for tab completion (e.g. `completion_for=myapp`), which generates the folder `completion` with `myapp.bash`, `myapp.zsh` and so forth.

**Comment by bitprophet** [comment]
Thanks for the feedback / diff!

I don't have time to rummage around in Program right now, but I thought it already had some field(s) for "what's the name you're usually bound to on the CLI?" - if it does, or if it doesn't and we add one, then yea, a `--print-completion-script` or whatnot option might be a semi easy one-shot.

**Comment by nhoening** [comment]
I hope I can find some time to look into it! (but the next weeks are packed)

You are correct, `Program` has a property called `binary`. And `--print-completion-script` is a better name then the one I proposed.

**Comment by nhoening** [comment]
@bitprophet I think the PR is now ready to be reviewed/merged. Let me know what I could do in the case it needs improvement.

**Comment by nhoening** [comment]
@bitprophet it seems you are finding some time for invoke at the moment. Any chance [my PR for this](https://github.com/pyinvoke/invoke/pull/414) can be merged in the near future? I can live with my workaround, but to have this supported natively would be nicer. I noticed there is a conflict now but it only concerns two lines of comments.

**Comment by bitprophet** [comment]
Thanks for keeping an eye on related tickets - at the moment I've got other higher priorities so not too worried about parallel work on #339 :) but good looking out! Will respond to the PR in its own ticket.

**Comment by bitprophet** [comment]
Merged #414! FWIW the changes are largely orthogonal re: #339, tho now that there's some light templating in the scripts, maybe that can impact it, not sure.
